cns/restserver/util.go

```diff
@@ -923,3 +923,60 @@ func (service *HTTPRestService) createNetworkContainers(createNetworkContainerRe
 		Message:    "",
 	}
 }
+
+// setResponse encodes the http response
+func (service *HTTPRestService) setResponse(w http.ResponseWriter, returnCode types.ResponseCode, response interface{}) {
+	serviceErr := service.Listener.Encode(w, &response)
+	logger.Response(service.Name, response, returnCode, serviceErr)
+}
+
+// generateAndCacheGetHomeAzResponse build the GetHomeAzResponse and cache it
+func (service *HTTPRestService) generateAndCacheGetHomeAzResponse(returnMessage string, returnCode types.ResponseCode, homeAzResponse nmagent.HomeAzResponse) cns.GetHomeAzResponse {
+	getHomeAzResponse := cns.GetHomeAzResponse{
+		HomeAzResponse: homeAzResponse,
+		Response:       cns.Response{ReturnCode: returnCode, Message: returnMessage},
+	}
+	service.updateGetHomeAzResponseCache(getHomeAzResponse)
+	return getHomeAzResponse
+}
+
+// readNMASupportedApisCache gets the nmsSupportedApisCache value
+func (service *HTTPRestService) readNMASupportedApisCache() []string {
+	service.RLock()
+	supportedApis := service.nmaSupportedApisCache
+	service.RUnlock()
+	return supportedApis
+}
```

Comment:
I know that doing this with a map would **not** be thread-safe; I am not sure if returning a slice like this is thread-safe, since the slice is a sort of reference type and the underlying array can be mutated through any copy of the slice?

Reply:
Ya, the slice is a sort of reference type.  I don't see any scenario we would mutate the copy slice for now. But I can do a deep copy here so it has 0 chance to mutate the underlying cache value.